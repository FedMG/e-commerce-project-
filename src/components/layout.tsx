import Head from "next/head";
import { RootLayout } from '../../additional'

import { Header } from "./structure/Header";

export const Layout: React.FC<RootLayout> = ({ children, title }) => {
  return (
    <>
      <Head>
        <title>{title}</title>
        <meta name="description" content="An E-commerce project" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <Header />
      <main>{children}</main>        
    </>
  );
};
