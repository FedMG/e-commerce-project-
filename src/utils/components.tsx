import { ReactElement } from 'react'
import { LinkProps } from 'additional'

export const Anchor: React.FC<LinkProps> = ({ path, route }): ReactElement => (
  <a
    href={path.length > 0 ? path : '#'}
    className='text-white hover:text-gray-400'
    target='_blank'
    rel='noreferrer'
  >
    {route}
  </a>
)
