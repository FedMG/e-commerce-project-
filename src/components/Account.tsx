import { useSession, signIn } from 'next-auth/react'
import Image from 'next/image'

import { List } from 'src/utils/components'
import { Dropdown } from './Dropdown'

const signUp = (): void => {}

const sessionButtons = [
  {
    route: 'Sign in',
    onClick: async () => await signIn()
  },
  {
    route: 'Sign up',
    onClick: () => signUp()
  }
]

const temp = {
  options: ['Profile', 'Your purchases', 'Your sales', 'Settings', 'Sign Out'],
  path: ['/profile', '/purchases', '/sales', '/settings', '/signout']
}

export const Account = (): JSX.Element | null => {
  const { data: session, status } = useSession()

  if (status === 'authenticated' && ((session?.user) != null)) {
    const { user } = session

    return (
      <>
        <Dropdown
          label={user.name ?? 'User account'}
          options={temp.options}
          path={temp.path}
        >
          {user.image && <Image src={user?.image} className='rounded-full' width='30' height='30' alt='User image profile' />}
        </Dropdown>
      </>
    )
  }

  if (session == null) {
    return (
      <>
        {sessionButtons.map(({ route, onClick }) => (
          <List key={route}>
            <button onClick={onClick}>{route}</button>
          </List>
        ))}
      </>
    )
  }
  return null
}
