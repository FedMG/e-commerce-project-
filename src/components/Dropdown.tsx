'use client'

import Link from 'next/link'
import { useState, ReactElement } from 'react'

import { DropdownProps } from 'additional'
import { Button } from 'src/utils/components'

const DropdownSVG = (): ReactElement => (
  <svg className='-mr-1 ml-2 h-5 w-5' viewBox='0 0 20 20' fill='currentColor'>
    <path
      fillRule='evenodd'
      d='M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z'
      clipRule='evenodd'
    />
  </svg>
)

export const Dropdown: React.FC<DropdownProps> = ({ children, label, options, path }): ReactElement => {
  const [isOpen, setIsOpen] = useState(false)

  return (
    <div className='relative inline-block text-left'>
      <Button
        style='inline-flex justify-center w-full text-sm leading-5 font-medium'
        onClick={() => setIsOpen(!isOpen)}
      >
        {label}
        {children ?? <DropdownSVG />}
      </Button>
      {isOpen && (
        <div className='origin-top-left absolute left-0 mt-2 w-48 rounded-md bg-white shadow-xs'>
          {options.map((option, i) => (
            <Link
              key={option}
              href={path[i]}
              className='block px-4 py-2 text-sm leading-5 text-gray-700 hover:bg-gray-100 rounded-md focus:outline-none focus:bg-gray-100'
              onClick={() => setIsOpen(false)}
            >
              {option}
            </Link>
          ))}
        </div>
      )}
    </div>
  )
}
