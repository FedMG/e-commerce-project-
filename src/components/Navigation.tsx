import Link from 'next/link'
import Image from 'next/image'
import { NavLinks } from 'additional'

import { List } from 'src/utils/components'
import { Dropdown } from './Dropdown'
import { Account } from './Account'

const navLinks: NavLinks[] = [
  {
    normal: {
      route: 'Home',
      path: '/'
    }
  },
  {
    normal: {
      route: 'About',
      path: '/about'
    }
  },
  {
    dropdown: {
      route: {
        label: 'Categories',
        options: ['Electronics', 'Clothes', 'Cars']
      },
      path: ['/categories', '/clothes', '/cars']
    }
  },
  {
    account: true
  },
  {
    special: {
      route: 'Add to Cart',
      path: '/cart',
      src: '/cart.svg'
    }
  }
]

export const Navigation = () => {
  return (
    <nav className='sticky top-0 max-w-full px-4 py-3 md:px-6 z-50 bg-[#d2ced5]'>
      <ul className='flex items-center justify-start mt-0 mr-6 space-x-8 text-sm font-medium'>
        {navLinks.map(({ normal, dropdown, account, special }) => {
          if (normal != null) {
            return (
              <List key={normal.route}>
                <Link href={normal.path}>{normal.route}</Link>
              </List>
            )
          }

          if (dropdown != null) {
            return (
              <List key={dropdown.route.label} radical='flex-auto'>
                <Dropdown
                  label={dropdown.route.label}
                  options={dropdown.route.options}
                  path={dropdown.path}
                />
              </List>
            )
          }

          if (account) {
            return <Account key='account' />
          }

          if (special != null) {
            return (
              <List key={special.route}>
                <Link href={special.path}>
                  <Image
                    src={special.src}
                    width={25}
                    height={25}
                    alt={`${special.route} icon`}
                  />
                </Link>
              </List>
            )
          }
        })}
      </ul>
    </nav>
  )
}
