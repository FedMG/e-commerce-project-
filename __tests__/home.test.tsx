import { render } from '@testing-library/react'

import { Layout } from 'src/components/layout'
import Home from 'src/pages'

it('should renders an home with a layout unchanged', () => {
  const { container } = render(<Layout title='Name - Home'><Home /></Layout>)
  expect(container).toMatchSnapshot()
})
