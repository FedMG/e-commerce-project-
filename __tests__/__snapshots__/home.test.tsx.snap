// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should renders an header component unchanged 1`] = `
<div>
  <header
    class="flex bg-orange-300 border-gray-200"
  >
    <div
      class="flex items-center px-2 py-2"
    >
      <img
        alt="E-commerce logo"
        class="mr-3"
        data-nimg="1"
        decoding="async"
        height="35"
        loading="lazy"
        src="/logo.svg"
        style="color: transparent;"
        width="35"
      />
      <span
        class="text-xl font-semibold"
      >
        E-commerce
      </span>
    </div>
  </header>
  <main>
    <h1
      class="text-3xl font-bold"
    >
      Home
    </h1>
  </main>
  <footer
    class="fixed bottom-0 w-full bg-gray-900 text-white py-6"
  >
    <div
      class="container mx-auto px-4"
    >
      <div
        class="flex items-center justify-between"
      >
        <div
          class="text-sm"
        >
          <p>
            Â© 2023 e-commerce
          </p>
        </div>
        <div
          class="flex flex-row sm:space-x-6"
        >
          <div>
            <h4>
              Policy
            </h4>
            <div
              class="flex flex-col"
            >
              <a
                class="text-white hover:text-gray-400"
                href="/privacy-policy"
              >
                Privacy Policy
              </a>
              <a
                class="text-white hover:text-gray-400"
                href="/terms-of-service"
              >
                Terms of Service
              </a>
              <a
                class="text-white hover:text-gray-400"
                href="/consumer-rights"
              >
                Consumer Rights
              </a>
            </div>
          </div>
          <div>
            <h4>
              Support
            </h4>
            <div
              class="flex flex-col"
            >
              <a
                class="text-white hover:text-gray-400"
                href="/contact-us"
              >
                Contact Us
              </a>
              <a
                class="text-white hover:text-gray-400"
                href="/forum"
              >
                Forum
              </a>
              <a
                class="text-white hover:text-gray-400"
                href="/common-questions"
              >
                Common questions
              </a>
            </div>
          </div>
          <div>
            <h4>
              Networks
            </h4>
            <div
              class="flex flex-col"
            >
              <a
                class="text-white hover:text-gray-400"
                href="/instagram/ecommerce"
                rel="noreferrer"
                target="_blank"
              >
                Instagram
              </a>
              <a
                class="text-white hover:text-gray-400"
                href="/twitter/ecommerce"
                rel="noreferrer"
                target="_blank"
              >
                Twitter
              </a>
              <a
                class="text-white hover:text-gray-400"
                href="/facebook/ecommerce"
                rel="noreferrer"
                target="_blank"
              >
                Facebook
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </footer>
</div>
`;
